File: awesomestuff/src/main/java/com/awesome/dialog/AwesomeSaveStateHandler.java
package com.awesome.dialog;

import android.os.Bundle;
import android.util.SparseArray;

import java.lang.ref.WeakReference;

/**
 * Created by yarolegovich on 17.04.2016.
 */
public class AwesomeSaveStateHandler {

    private static final String KEY_DIALOG_ID = "id";

    private SparseArray<WeakReference<AwesomeBase<?>>> handledDialogs;

    public AwesomeSaveStateHandler() {
        handledDialogs = new SparseArray<>();
    }

    public void saveInstanceState(Bundle outState) {
        for (int index = handledDialogs.size() - 1; index >= 0; index--) {
            WeakReference<AwesomeBase<?>> dialogRef = handledDialogs.valueAt(index);
            if (dialogRef.get() == null) {
                handledDialogs.remove(index);
                continue;
            }
            AwesomeBase<?> dialog = dialogRef.get();
            if (dialog.isShowing()) {
                dialog.onSaveInstanceState(outState);
                outState.putInt(KEY_DIALOG_ID, handledDialogs.keyAt(index));
                return;
            }
        }
    }

    void handleDialogStateSave(int id, AwesomeBase<?> dialog) {
        handledDialogs.put(id, new WeakReference<AwesomeBase<?>>(dialog));
    }

    public static boolean wasDialogOnScreen(Bundle savedInstanceState) {
        return savedInstanceState.keySet().contains(KEY_DIALOG_ID);
    }

    public static int getSavedDialogId(Bundle savedInstanceState) {
        return savedInstanceState.getInt(KEY_DIALOG_ID, -1);
    }
}

File: fusion/src/main/java/com/awesome/dialog/AwesomeSaveStateHandler.kt
package com.awesome.dialog

import android.os.Bundle
import android.util.SparseArray

import java.lang.ref.WeakReference

/**
 * Created by Sandip Savaliya on 17.04.2016.
 */
class AwesomeSaveStateHandler {

    private val handledDialogs: SparseArray<WeakReference<AwesomeBase<*>>>

    init {
        handledDialogs = SparseArray()
    }

    fun saveInstanceState(outState: Bundle) {
        for (index in handledDialogs.size() - 1 downTo 0) {
            val dialogRef = handledDialogs.valueAt(index)
            if (dialogRef.get() == null) {
                handledDialogs.remove(index)
                continue
            }
            val dialog = dialogRef.get()
            if (dialog!!.isShowing) {
                dialog.onSaveInstanceState(outState)
                outState.putInt(KEY_DIALOG_ID, handledDialogs.keyAt(index))
                return
            }
        }
    }

    internal fun handleDialogStateSave(id: Int, dialog: AwesomeBase<*>) {
        handledDialogs.put(id, WeakReference(dialog))
    }

    companion object {

        private val KEY_DIALOG_ID = "id"

        fun wasDialogOnScreen(savedInstanceState: Bundle): Boolean {
            return savedInstanceState.keySet().contains(KEY_DIALOG_ID)
        }

        fun getSavedDialogId(savedInstanceState: Bundle): Int {
            return savedInstanceState.getInt(KEY_DIALOG_ID, -1)
        }
    }
}
